### PR TITLE
feat: Add translation-cursor-deploy

### PR DESCRIPTION
**Pull request by igorDolzh**
https://linear.app/pleo/issue/HMS-373/create-cdn-solution-for-translation-files-for-product-web

This PR is about creating the logic for updating translation cursor deploy files.

There are two files `/latest` and `/previous`. 
`/latest` contains latest hash for translation files which will be used for all S3 buckets and environments.
`/previous` contains previous hash for translation files. And it is needed for "fast" rollback, when the user doesn't provide exact hash on what he wants to update

More info could be found in notion page:

https://www.notion.so/pleo/Migration-translation-files-from-Github-to-S3-for-product-web-for-now-e8e54d04398342ecab305a43ab6ebe6a for `Rollback` and `Translation cursor files` sections

### CODE REVIEW

File: actions/translation-cursor-deploy/action.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+name: 'Translation Deploy Action'
+description:
+    'Deploy the new version of the translation files by modifying a cursor file in S3 bucket.'
+author: 'Pleo'
+inputs:
+    hash:
+        description: Hash which is equal to run id
```
Comment (skovhus) [inline actions/translation-cursor-deploy/action.yml:7]: What does `run id` mean?

File: actions/translation-cursor-deploy/main.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/**
+ * Custom Cursor Deploy GitHub action
```
Comment (skovhus) [inline actions/translation-cursor-deploy/main.ts:2]: We should have a test here documenting the behaviour and all the corner cases. You should be able to reuse the setup from other actions.

File: actions/translation-cursor-deploy/main.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    const bucket = core.getInput('bucket_name', {required: true})
+    const hash = core.getInput('hash', {required: true})
+
+    await cursorDeploy({
+        bucket,
+        hash
+    })
+})
+
+const PREVIOUS_HASH_COMMAND = 'previous' // special command for rollback to hash from /previous
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:23]: This seems quite hacky to me (i.e. reusing hash to also double as a command sometimes). Maybe we can use the same concept of "deploy mode" we have in the regular deploy action, this would make things much more explicit.

File: actions/translation-cursor-deploy/main.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            core.setFailed('Previous cursor is empty')
+            return
+        }
+
+        const isLatestFileExists = await fileExistsInS3({bucket, key: 'translation-deploy/latest'})
+        if (isLatestFileExists) {
+            await exec('aws s3 rm ', [latestPath])
+        }
+        await exec('aws s3 cp ', [previousPath, latestPath])
+    } else {
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:49]: I suggest an early return in the `if` case, instead of using `else` - this would make the default code path more pronounced.

File: actions/translation-cursor-deploy/main.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+const PREVIOUS_HASH_COMMAND = 'previous' // special command for rollback to hash from /previous
+
+interface CursorDeployActionArgs {
+    bucket: string
+    hash: string
+}
+
+export async function cursorDeploy({bucket, hash}: CursorDeployActionArgs) {
+    const previousPath = `s3://${bucket}/translation-deploy/previous`
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:31]: Should we extract `translation-deploy/previous` and `translation-deploy/latest`  to constants?

File: actions/translation-cursor-deploy/main.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        if (!isPreviosFileExists) {
+            core.setFailed('Previous cursor is empty')
+            return
+        }
+
+        const isLatestFileExists = await fileExistsInS3({bucket, key: 'translation-deploy/latest'})
+        if (isLatestFileExists) {
+            await exec('aws s3 rm ', [latestPath])
+        }
+        await exec('aws s3 cp ', [previousPath, latestPath])
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:48]: I suggest using the helper method for copying to S3 https://github.com/pleo-oss/pleo-spa-cicd/blob/main/actions/utils.ts#L64 - it will make it easier to mock during testing.

File: actions/translation-cursor-deploy/main.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            key: 'translation-deploy/previous'
+        })
+        if (!isPreviosFileExists) {
+            core.setFailed('Previous cursor is empty')
+            return
+        }
+
+        const isLatestFileExists = await fileExistsInS3({bucket, key: 'translation-deploy/latest'})
+        if (isLatestFileExists) {
+            await exec('aws s3 rm ', [latestPath])
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:46]: So we're deleting the latest file here, and I'm guessing there will be a moment where a request to get the latest file would fail (i.e. before we upload the new one). Why do we need to delete and upload, instead of replacing the file?

File: actions/translation-cursor-deploy/main.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        await exec('aws s3 cp ', [previousPath, latestPath])
+    } else {
+        const isPreviosFileExists = await fileExistsInS3({
+            bucket,
+            key: 'translation-deploy/previous'
+        })
+        if (isPreviosFileExists) {
```
Comment (pekala) [inline actions/translation-cursor-deploy/main.ts:54]: Same here for replacing, instead of deleting and then copying

File: actions/translation-cursor-deploy/main.ts (hunk 3 above)
Comment (igorDolzh) [reply actions/translation-cursor-deploy/main.ts:23]: Could elaborate about concept of "deploy mode" ?

File: actions/utils.ts [hunk 9]
```diff
@@ -150,3 +150,9 @@ export async function getTreeHashForCommitHash(commit: string) {
 export async function getCurrentRepoTreeHash() {
     return getTreeHashForCommitHash('HEAD')
 }
+
+/**
+ * Keys for 'latest' and 'previous' cursor for translation files
+ */
+export const previousKey = `translation-deploy/previous`
+export const latestKey = `translation-deploy/latest`
```
Comment (pekala) [inline actions/utils.ts:158]: Any reason this needs to be in the shared utils? Don't think this is shared

**Review by pekala — COMMENTED** [review]
As discussed, I don't think we will need this if we adjust our regular deployment action a bit. I can have a stab at doing that.

**Comment by igorDolzh** [comment]
@pekala We agreed that we will do adjustments after I ship Translation CI/CD.

Despite this adjustment will be relatively not complex, it shift the focus out of the Translation CI/CD, and definitely not a blocker.

Let's merge this one and later PRs and then we may do such adjustment

**Review by pekala — APPROVED** [review]
Since we agree this is throw-away code, I won't be reviewing to keep it on the quality level expected here (e.g.  documentation) - but let's follow up with getting rid of it quickly.
